### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by TheVovchik**
**[DEMO LINK](https://thevovchik.github.io/react_movies-list-fetch-movies/)**

### CODE REVIEW

**Comment by ihorandrianov** [comment]
Add warning when trying to add duplicate movie!

File: src/api.ts [hunk 1]
```diff
@@ -1,13 +1,13 @@
 import { MovieData } from './types/MovieData';
 import { ResponseError } from './types/ReponseError';
 
-const API_URL = 'https://www.omdbapi.com/?apikey=your-key';
+const API_URL = 'https://www.omdbapi.com/?apikey=378681db';
 
 export function getMovie(query: string): Promise<MovieData | ResponseError> {
```
Comment (ihorandrianov) [inline src/api.ts:6]: Some mentors asks to rewrite this function with async syntax

File: src/components/Button/Button.tsx [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import { FC } from 'react';
+
+type Props = React.ButtonHTMLAttributes<HTMLButtonElement> & {
+  children: React.ReactNode;
+};
+
+export const Button: FC<Props> = ({ children, ...props }) => {
```
Comment (ihorandrianov) [inline src/components/Button/Button.tsx:7]: Why you need Button wrapper?

File: src/components/FindMovie/FindMovie.tsx [hunk 3]
```diff
@@ -1,10 +1,95 @@
...
+  const checkMovie = (newMovie: Movie) => {
+    return movies.some(film => film.imdbId === newMovie.imdbId);
+  };
+
+  const getMovieFromServer = async (appliedQuery: string) => {
+    const newFilm = await getMovie(appliedQuery);
+
+    setIsSearch(false);
+
+    if (Object.prototype.hasOwnProperty.call(newFilm, 'Poster')) {
```
Comment (ihorandrianov) [inline src/components/FindMovie/FindMovie.tsx:44]: You can simplify this syntax

File: src/components/AppProvider/AppProvider.tsx [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+});
+
+export const AppProvider: React.FC = ({ children }) => {
+  const [movies, setMovies] = useState<Movie[]>([]);
+  const [query, setQuery] = useState('');
+  const [movie, setMovie] = useState<Movie | null>(null);
+  const [isMovieExist, setIsMovieExist] = useState(false);
+  const [isSearch, setIsSearch] = useState(false);
+
+  const contextValue = useMemo(() => ({
```
Comment (ihorandrianov) [inline src/components/AppProvider/AppProvider.tsx:37]: Using useMemo without good reason is bad practice, overall this is strange use case to memoize function that did not calculate anything and just returns object of const values

File: src/components/Button/Button.tsx (hunk 2 above)
Comment (TheVovchik) [reply src/components/Button/Button.tsx:7]: simple practice to move some common logic

**Review by ihorandrianov — APPROVED** [review]

File: src/api.ts [hunk 5]
```diff
@@ -1,13 +1,13 @@
...
   return fetch(`${API_URL}&t=${query}`)
     .then(res => res.json())
-    .catch(() => ({
+    .catch((error) => ({
       Response: 'False',
-      Error: 'unexpected error',
+      Error: error,
```
Comment (Anna-Tkachenko) [inline src/api.ts:11]: you can also throw new error here and handle this error in catch {} while calling api function

File: src/components/AppProvider/AppProvider.tsx [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  setIsMovieExist: (newCheck: boolean) => void,
+  isSearch: boolean,
+  setIsSearch: (newStatus: boolean) => void,
+};
+
+export const AppContext = React.createContext<Context>({
+  movies: [],
+  setMovies: () => {},
+  query: '',
+  setQuery: () => {},
```
Comment (Anna-Tkachenko) [inline src/components/AppProvider/AppProvider.tsx:21]: you shouldn't add variables that can be changes on each click to the context
use context only for general values for all App, that can affect on many components and will use everywhere

File: src/components/Button/Button.tsx (hunk 2 above)
Comment (Anna-Tkachenko) [reply src/components/Button/Button.tsx:7]: good job with separate component for buttons!

**Review by Anna-Tkachenko — APPROVED** [review]
Good job!

**Pull request closed without merging**